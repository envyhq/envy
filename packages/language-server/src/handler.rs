use std::{error::Error, fmt};

use crate::{
    store::FileStore,
    types::{Initialize, LspNotification, LspRequest, TextDocumentHover},
    visitor::LspMessage,
};
use nv_lexer::TokenPosition;
use nv_parser::{AbstractSyntaxNode, DeclarationNode};
use nv_var_resolver::TreeResolver;
use serde_json::{json, Value};
use tokio::io::AsyncWriteExt;
use tokio::net::UnixStream;

pub enum TextDocumentSyncKind {
    Incremental = 2,
}

pub trait Handler<T> {
    async fn handle(&mut self, request: T) -> Result<Value, anyhow::Error>;
}

pub struct InitializeHandler;
impl Handler<Initialize> for InitializeHandler {
    async fn handle(&mut self, req: Initialize) -> Result<Value, anyhow::Error> {
        Ok(json!({
            "jsonrpc": "2.0",
            "id": req.base.id,
            "result": {
                "capabilities": {
                    "textDocumentSync": TextDocumentSyncKind::Incremental as i64,
                    "hoverProvider": true
                }
            }
        }))
    }
}

pub struct TextDocumentHoverHandler {
    pub file_path: String,
    pub file_store: FileStore,
}

impl Handler<TextDocumentHover> for TextDocumentHoverHandler {
    async fn handle(&mut self, req: TextDocumentHover) -> Result<Value, anyhow::Error> {
        let file = self.file_store.get(&self.file_path).unwrap();

        let cursor_position = TokenPosition {
            line: req.params.position.line as usize + 1,
            column: req.params.position.character as usize + 1,
        };

        let node = file.position_indexer.node_at(&cursor_position);

        if let Some(node) = node {
            if let Some(node) = node.upgrade() {
                let resolve_node =
                    AbstractSyntaxNode::Declaration(DeclarationNode::VarDeclaration(node));
                let resolved = file.resolver.resolve(&resolve_node).await?;
                let resolved = resolved.first().unwrap();

                return Ok(json!({
                    "jsonrpc": "2.0",
                    "id": req.base.id,
                    "result": {
                        "contents": format!("|Provider|Key|Value|
|---|---|---|
|{}|{}|{}|", resolved.provider, resolved.key, resolved.value.clone().unwrap_or("Not set".to_owned()))

                }
                }));
            }
        }

        log::debug!("Hover found no node");

        Ok(json!({
            "jsonrpc": "2.0",
            "id": req.base.id,
            "result": null
        }))
    }
}

#[derive(Debug, Clone)]
pub struct LspRequestError;

impl Error for LspRequestError {}

impl fmt::Display for LspRequestError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "invalid first item to double")
    }
}

pub struct LspRequestHandler {
    pub file_store: FileStore,
}
impl LspRequestHandler {
    async fn handle(&self, req: LspRequest) -> Result<Value, anyhow::Error> {
        let result = match req {
            LspRequest::Initialize(req) => InitializeHandler.handle(req).await?,
            LspRequest::TextDocumentHover(req) => {
                TextDocumentHoverHandler {
                    file_path: req.clone().params.text_document.uri.replace("file://", ""),
                    file_store: self.file_store.clone(),
                }
                .handle(req)
                .await?
            }
            LspRequest::Shutdown(_req) => {
                std::process::exit(0);
            }
        };

        Ok(result)
    }
}

pub struct LspNotificationHandler;
impl LspNotificationHandler {
    async fn handle(&self, notif: LspNotification) {
        match notif {
            LspNotification::Initialized(_params) => {
                log::info!("nv language server client connection initialized");
            }
            _ => {
                log::warn!("unhandled notification");
            } //
              // LspNotification::TextDocumentDidOpen(_params) => {}
              // LspNotification::TextDocumentDidClose(_params) => {}
              // LspNotification::TextDocumentDidSave(_params) => {}
              // LspNotification::TextDocumentDidChange(_params) => {}
        }
    }
}

pub struct LspMessageHandler;
impl LspMessageHandler {
    pub async fn handle(
        &self,
        socket: &mut UnixStream,
        message: LspMessage,
        file_store: FileStore,
    ) -> Result<(), anyhow::Error> {
        match message {
            LspMessage::Request(req) => {
                let result = LspRequestHandler { file_store }.handle(req).await?;
                let result = serde_json::to_string(&result)?;
                let len = result.as_bytes().len();

                let response = format!("Content-Length: {}\r\n\r\n{}", len, result);

                socket.write_all(response.as_bytes()).await?;
            }
            LspMessage::Notification(notif) => LspNotificationHandler.handle(notif).await,
        }

        Ok(())
    }
}
