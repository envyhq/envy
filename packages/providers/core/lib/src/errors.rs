use std::fmt::Debug;

#[derive(Debug)]
pub enum ServerError {
    Write,
    Read,
    NoMessages,
    SocketError,
    // TODO: not error?
    WouldBlock,
}

#[derive(Debug)]
pub enum ServerResult {
    ClientDisconnected,
    CancelledGracefully,
}

#[derive(Debug, Clone)]
pub struct ProviderRetrievalError {
    pub message: Option<String>,
}

#[derive(Debug, Clone)]
pub enum ProviderError {
    NoValueForKey,
    RetrievalError(ProviderRetrievalError),
    ExplodeyProvider,
}
