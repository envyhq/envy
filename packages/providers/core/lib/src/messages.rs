pub type Message = Vec<u8>;
