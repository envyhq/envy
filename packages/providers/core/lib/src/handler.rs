use crate::errors::{ServerError, ServerResult};
use crate::messages::Message;
use crate::protocol::MessageSerializer;
use crate::{Controller, ProviderError};
use std::fmt::Debug;
use std::io::ErrorKind;
use std::sync::Arc;
use tokio::net::UnixStream;
use tokio::sync::mpsc;
use tokio::task::JoinHandle;
use tokio_util::sync::CancellationToken;

#[derive(Debug)]
pub struct Handler {
    pub stream: Arc<UnixStream>,
    pub controller: Arc<dyn Controller>,
    pub serializer: MessageSerializer,
}

impl Handler {
    pub fn new(stream: Arc<UnixStream>, controller: Arc<dyn Controller>) -> Self {
        Handler {
            stream,
            controller,
            serializer: MessageSerializer {},
        }
    }

    // In parallel, we read and write from the full duplex stream when its ready for such op.
    // "Actions" are also processed in their own green thread to allow for actions of varying
    // durations to be processed concurrently without blocking queue I/O.
    pub async fn handle(self: Arc<Self>) -> Result<ServerResult, ServerError> {
        // TODO: think about whether or not we need to authenticate/authorize
        // new clients in any way. It will probably go here.
        // Had a thought today that the root problem is how to share a secret with both
        // the client app and the nv unix system. I thought about setting at compile/build time for
        // both but I feel like its essentially the same as saving to FS just w/ obfuscation.
        // We need a dynamic way to auth a client with server, that the server is actually the
        // inteded app. Maybe look into code signing for inspo?
        // .......
        // Lots more thinking today. Another idea...
        // We leave most of the auth to the providers. e.g. locally setting AWS creds in env vars
        // and conforming to AWS idea of identity on their machines.
        // Another potential to maybe harden is to write some secret token at startup or build time
        // that is consumed by the server and client once. This means only the first claim of the
        // container will have access for the rest of the duration of the container?
        // Also, we could ignore all the above and have some other agent/server that is
        // completely separate and private. This would essentially be stepping into the role of the providers
        // to some degree. For example, hashicorpt vault is secret management server that auths and comms with clients.

        let (req_tx, req_rx) = mpsc::channel(16);
        let (res_tx, res_rx) = mpsc::channel(16);
        let cancel_token = CancellationToken::new();

        let read_out = req_tx.clone();
        let read_stream = self.stream.clone();
        let read_cancel = cancel_token.clone();
        let read_thread: JoinHandle<Result<ServerResult, ServerError>> = tokio::spawn(async move {
            tokio::select! {
                _ = read_cancel.cancelled() => {
                    Ok(ServerResult::ClientDisconnected)
                }
                result = async {
                    loop {
                        let ready = read_stream.readable().await;

                        match ready {
                            Ok(_) => {
                                let request = Handler::read_message(read_stream.clone()).await;

                                log::debug!("Read message; request = {:?}", request);

                                match request {
                                    Ok(request) if !request.is_empty() => {
                                        let result = read_out.send(request.clone()).await;

                                        if let Err(e) = result {
                                            log::error!("{:?} Failed to send out read response; err = {:?}", String::from_utf8(request), e);
                                            break Err(ServerError::SocketError);
                                        }
                                    }
                                    Ok(_) => {
                                        // Empty message means client shutdown
                                        break Ok(ServerResult::ClientDisconnected);
                                    },
                                     Err(ServerError::WouldBlock) => {
                                        continue;
                                    }
                                    Err(e) => {
                                        log::error!("Failed to read message; err = {:?}", e);
                                        break Err(ServerError::SocketError);
                                    }
                                }
                            }
                            Err(e) => {
                                log::error!("Failed to read readiness; err = {:?}", e);
                                break Err(ServerError::SocketError);
                            }
                        }
                    }
                } => {
                    result
                }
            }
        });

        let mut action_in = req_rx;
        let action_out = res_tx.clone();
        let controller = self.controller.clone();
        let action_cancel = cancel_token.clone();
        let action_thread: JoinHandle<Result<ServerResult, ServerError>> = tokio::spawn(
            async move {
                tokio::select! {
                    _ = action_cancel.cancelled() => {
                        Ok(ServerResult::CancelledGracefully)
                    }
                    result = async {
                        let res: Result<ServerResult, ServerError> = loop {
                            let message = action_in.recv().await;
                            if message.is_none() {
                                continue;
                            }

                            let action = message.unwrap();

                            let response = controller.action(&action).await;

                            let result = action_out.send(response).await;

                            if result.is_err() {
                                log::error!("Failed to action message; action = {:?} err = {:?}", String::from_utf8(action), result);
                                break Err(ServerError::SocketError);
                            }
                        };

                        res
                    } => {
                        result
                    }
                }
            },
        );

        let write_stream = self.stream.clone();
        let mut write_in = res_rx;
        let action_out_recover = res_tx;
        let write_cancel = cancel_token.clone();
        let handler: Arc<Handler> = Arc::clone(&self);
        let write_thread: JoinHandle<Result<ServerResult, ServerError>> = tokio::spawn(
            async move {
                tokio::select! {
                    _ = write_cancel.cancelled() => {
                        Ok(ServerResult::CancelledGracefully)
                    }
                    result = async {
                        loop {
                            let message = write_in.recv().await;

                            if message.is_none() {
                                continue;
                            }

                            let message = message.unwrap();


                            let ready = write_stream.writable().await;

                            match ready {
                                Ok(_) => {
                                    let result = handler.write_message(write_stream.clone(), message.clone()).await;

                                        match result {
                                            Err(ServerError::WouldBlock) => {
                                                // TODO: test that this is required
                                                let _ = action_out_recover.send(message).await;
                                                continue;
                                            }
                                            Err(e) => {
                                                log::error!("Failed to write response; err = {:?}", e);
                                                break Err(e)
                                            }
                                            Ok(_) => { continue; }
                                        }
                                }
                                Err(e) => {
                                    log::error!("Failed to read readiness on write; err = {:?}", e);
                                    break Err(ServerError::SocketError)
                                }
                            }
                        }
                    } => {
                        result
                    }

                }
            },
        );

        tokio::select! {
                result = read_thread => {
                    log::debug!("Read thread cancelled; result = {:?}", result);
                    cancel_token.cancel();
                    result.unwrap_or(Err(ServerError::SocketError))
                }
                result = action_thread => {
                    log::debug!("Action thread cancelled; result = {:?}", result);
                    cancel_token.cancel();
                    result.unwrap_or(Err(ServerError::SocketError))
                }
                result = write_thread => {
                    log::debug!("Write thread cancelled; result = {:?}", result);
                    cancel_token.cancel();
                    result.unwrap_or(Err(ServerError::SocketError))
                }
        }
    }

    async fn read_message(stream: Arc<UnixStream>) -> Result<Message, ServerError> {
        let mut message = vec![0; 1024];
        match stream.try_read(&mut message) {
            Ok(bytes_read) => {
                message.truncate(bytes_read);

                Ok(message)
            }
            Err(ref e) if e.kind() == ErrorKind::WouldBlock => Err(ServerError::WouldBlock),
            Err(e) => {
                log::error!("Failed to read from socket; err = {:?}", e);
                Err(ServerError::Read)
            }
        }
    }

    async fn write_message(
        &self,
        stream: Arc<UnixStream>,
        res: Result<Message, ProviderError>,
    ) -> Result<(), ServerError> {
        let message = res.unwrap_or("nout".as_bytes().into());
        let serialized = self.serializer.serialize(message);

        match stream.try_write(&serialized) {
            Ok(_) => Ok(()),
            Err(ref e) if e.kind() == ErrorKind::WouldBlock => Err(ServerError::Write),
            Err(e) => {
                log::error!("Failed to write to socket; err = {:?}", e);
                Err(ServerError::Write)
            }
        }
    }
}
