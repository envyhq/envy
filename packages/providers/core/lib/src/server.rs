use crate::errors::ServerError;
use crate::handler::Handler;
use crate::messages::Message;
use crate::ProviderError;
use async_trait::async_trait;
use std::fmt::Debug;
use std::fs::remove_file;
use std::io::ErrorKind;
use std::process::exit;
use std::sync::Arc;
use tokio::net::UnixListener;

#[async_trait]
pub trait Controller: Sync + Send + Debug {
    async fn action(&self, message: &Message) -> Result<Message, ProviderError>;
}

#[derive(Debug)]
pub struct Server {
    pub path: String,
    pub controller: Arc<dyn Controller>,
}

impl Server {
    pub fn new(path: &str, controller: Arc<dyn Controller>) -> Self {
        Self {
            path: path.to_owned(),
            controller,
        }
    }

    pub async fn start(&self) -> Result<(), ServerError> {
        let path = self.path.clone();
        let clean_up = move || {
            let result = remove_file(path.clone());

            if let Err(error) = result {
                if error.kind() == ErrorKind::NotFound {
                    return;
                }

                log::error!("Error removing file: {:?}", error);
            }
        };

        clean_up();

        let listener = UnixListener::bind(self.path.clone()).unwrap();
        // TODO: move out of lib
        ctrlc::set_handler(move || {
            clean_up();
            exit(0);
        })
        .expect("Error setting Ctrl-C handler");

        loop {
            let socket = listener.accept().await;
            log::debug!("New client connected {:?}", socket);

            match socket {
                Ok((stream, _addr)) => {
                    let controller = self.controller.clone();
                    tokio::spawn(async move {
                        let stream = Arc::new(stream);
                        Handler::new(stream, controller).handle().await
                    });
                }
                Err(e) => {
                    log::error!("{:?}", e);
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Controller;
    use crate::ProviderError;
    use crate::{messages::Message, Server};
    use async_trait::async_trait;
    use std::sync::Arc;
    use std::time::Duration;
    use tokio::io::{AsyncReadExt, AsyncWriteExt};
    use tokio::net::UnixStream;
    use tokio::time::sleep;

    #[derive(Debug)]
    struct TestController {}

    #[async_trait]
    impl Controller for TestController {
        async fn action(&self, message: &Message) -> Result<Message, ProviderError> {
            Ok(message.clone())
        }
    }

    #[tokio::test]
    async fn test_server() {
        let controller = Arc::new(TestController {});
        let path = "/tmp/nv-provider.sock";

        let server = Server::new(path, controller);

        tokio::spawn(async move {
            let _ = server.start().await;
        });
        // TODO: no sleepy
        sleep(Duration::from_millis(100)).await;

        let mut client = UnixStream::connect(path).await.unwrap();
        let msg = b"who";
        client.write_all(msg).await.unwrap();

        let mut buf = [0; 1024];
        let n = client.read(&mut buf).await.unwrap();

        let response = String::from_utf8((buf[..n]).to_vec());

        assert_eq!(response, Ok("who".to_owned()));
    }
}
