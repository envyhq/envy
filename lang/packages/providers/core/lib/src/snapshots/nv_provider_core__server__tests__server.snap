---
source: packages/providers/core/lib/src/server.rs
expression: header.checksum
---
58602
